File: egs/openasr/local/make_trans.py
#!/usr/bin/python
# -*- encoding: utf-8 -*-
__author__ = 'feiteng'
# -*- coding: utf-8 -*-
import re
import sys
import random
import string
import os
import hashlib

def get_hash_str(val):
    return hashlib.md5(str(val)).hexdigest()

class MakeTrans:
    def __init__(self, postfix, in_flist, input_snr, out_txt, out_scp, utt2spk):
        self.SpknameDict = dict()
        self.TspDict = dict()

        self.F = open(input_snr, 'r').readlines()

        self.G = open(in_flist, 'r').readlines()
        self.postfix = postfix

        self.O = open(out_txt, 'w')
        self.P = open(out_scp, 'w')
        self.U = open(utt2spk, 'w')
        self.Fdict()

    def Fdict(self):
        for line in self.F:
            m = re.match(r'(.+) \((.+)\)$', line.strip())
            self.TspDict[m.group(2).strip()] = m.group(1).strip().lower()

    def ProcessAllInOne(self):
        no_tsp_num = 0
        utt2spk = dict()
        for i in range(0,self.G.__len__()):
            # 4584_engzo_2603_ios
            # 4294_engzo_1983_android
            # 4672_engzo_2577_android
            if not os.path.isfile(self.G[i].strip()):
                continue
            wavname = self.G[i].strip().split("/")[-1]
            sp = wavname.split("_")
            spkname = sp[0]

            if True:
                spkname = get_hash_str(spkname)
                assert spkname != ''
                if wavname.find('ios') > 0:
                    spkname = spkname
                else:
                    spkname = spkname
            else:
                continue

            if sp[1] != 'engzo' and not wechat:
                print self.G[i].strip(), " bad wav file name."
                continue

            uttname = sp[2]
            spkuttname = wavname[:-4]
            # print sp
            if len(sp[0]) == 0 or len(uttname)==0 or len(spkuttname) == 0:
                print self.G[i].strip(), " bad wav file name."
                continue

            if self.G[i].strip()[-4:] != ".wav":
                print self.G[i].strip(), " is not correct wav file name."
                continue

            if self.TspDict.has_key(spkuttname):
                key = spkname + '_' + spkuttname.replace('.', '_') + "_" + self.postfix
                if not utt2spk.has_key(spkname):
                    utt2spk[spkname] = []
                utt2spk[spkname].append(key)

                print >> self.O, "%s %s" %(key, self.TspDict[spkuttname])
                print >> self.P, "%s %s" %(key, self.G[i].strip())
            else:
                no_tsp_num += 1
                #print "%s no trans for sent utt: %s" %(self.postfix, spkuttname)

            if i % 100000 == 0 :
                print " make_trans.py has process %f" % (float(i)/self.G.__len__())
        for spkname in sorted(utt2spk.keys()):
            utts = utt2spk[spkname]
            assert len(utts) > 0
            for utt in utts:
                print >>self.U, '%s %s' % (utt, spkname)

        print "%s wavs have no transcriptions. percent %.3f" %(no_tsp_num, no_tsp_num*100.0/self.G.__len__())

        self.O.close()
        self.P.close()
        self.U.close()


if __name__ == "__main__":

    if len(sys.argv) == 8:
        wechat = (sys.argv[7] == 'True')
    print "Wechat =", wechat

    MakeTrans(sys.argv[1], sys.argv[2], sys.argv[3], sys.argv[4], sys.argv[5], sys.argv[6])
    print " Done. "

File: egs/openasr/local/make_trans.pl
#!/usr/bin/perl
# Copyright 2010-2012 Microsoft Corporation  Johns Hopkins University (Author: Daniel Povey)
# Feiteng 2014.12.23 12.25 简单修改

# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#  http://www.apache.org/licenses/LICENSE-2.0
#
# THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
# WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
# MERCHANTABLITY OR NON-INFRINGEMENT.
# See the Apache 2 License for the specific language governing permissions and
# limitations under the License.


# usage:  make_trans.sh postfix in.flist input.snr out.txt out.scp

# postfix is last letters of the database "key" (rest are numeric)

# in.flist is just a list of filenames, probably of .sph files.
# input.snr is an snr format file from the RM dataset.  
# out.txt is the output transcriptions in format "key word1 word\n"
# out.scp is the output scp file, which is as in.scp but has the
# database-key first on each line.

# Reads from first argument e.g. $rootdir/rm1_audio1/rm1/doc/al_sents.snr
# and second argument train_wav.scp 
# Writes to standard output trans.txt

use File::Basename;

if(@ARGV != 5) {
    die "usage:  make_trans.sh postfix in.flist input.snr out.txt out.scp\n";
}
($postfix, $in_flist, $input_snr, $out_txt, $out_scp) = @ARGV;

open(F, "<$input_snr") || die "Opening SNOR file $input_snr";

while(<F>) {
    if(m/^;/) { next; }
    m/(.+) \((.+)\)/ || die "bad line $_"; #这里根据$input_snr的格式做匹配
    $T{$2} = $1;
}

close(F);
open(G, "<$in_flist") || die "Opening file list $in_flist";

open(O, ">$out_txt") || die "Open output transcription file $out_txt";

open(P, ">$out_scp") || die "Open output scp file $out_scp";

while(<G>) {
    $_ =~ m:/(.+)/((\w+)_engzo_(\d+)_(\w+)(\.?)(\w+))\.wav$:i || die "bad wav line $_";
    $spkname = $3;
    $uttname = $4;
    $spkuttname = $2;

    #$uttname  =~ tr/a-z/A-Z/; #关闭大小写的转换
    defined $T{$spkuttname} || "no trans for sent $uttname" && next;
    $spkname =~ s/_//g; # remove underscore from spk name to make key nicer.
    $key = $spkname . "_" . $uttname . "_" . $postfix;
    #$key =~ tr/A-Z/a-z/; # Make it all lower case.
    # to make the numerical and string-sorted orders the same.
    # print "$key $T{$spkuttname}\n";
    # print "$key $_" || die "Error writing to wav file list";

    print O "$key $T{$spkuttname}\n";
    print P "$key $_" || die "Error writing to wav file list";
    $n++;
}
close(G) || die "Closing input.";
close(O) || die "Closing output.";
close(P) || die "Closing output.";
